### PR TITLE
Desktop: Fixes #5178: NoteListItem's height can be properly styled using CSS.

### LINKED ISSUE #5178: Incorrect number of note items and unused dead area in Note List panel
To increase the displayed information of Joplin, I reduced the height of note items in the Note List using the following userchrome.css. 

    .sidebar div.list-item-container {
        height: 20px !important;
    }

    .note-list .list-item-container,
    .note-list .list-item-container div {
        height: 25px !important;
    }

The result is as follows. <strong>The bottom area of the Note List is unused</strong>, and <strong>the number of the notes in the list is unchanged before and after the height is changed</strong>. For comparison, the Sidebar is properly rendered. 

joplin_bug_20210711_2a

For reference, the before image is here. 

joplin_bug_20210711_1

## Environment

Joplin version: 2.1.8
Platform: Windows
OS specifics: Windows 10 Pro (21H1)

## Steps to reproduce

1. Run with the above userchrome.css.
2. Click 'All notes' in the Sidebar. If you have enough number of notes, you always observe this bug in the Note List panel.

## Describe what you expected to happen

When the height of items in the Note List is reduced, the whole area should be used and the number of items should be increased as follows.

joplin_bug_20210711_expected

### PR DESCRIPTION
**Pull request by ken1kob**
The bug reported in issue #5178 is fixed.

When the height of NoteListItem is styled using `userchrome.css` as below, the number of displayed note items is correctly calcurated, and there is no unused dead area in the NoteList panel.

    .note-list .list-item-container,
    .note-list .list-item-container div {
        height: 25px !important;
    }

By applying this fix, you see what people expects as below shown in issue #5178.

joplin_bug_20210711_expected

In this case, React enables us to know the height of NoteListItem specified in CSS before the first time of the NoteList panel is displayed. 

This fix does not introduce any performance degradation. Both handling thousands of notes and the flexible styling using CSS go together.

### CODE REVIEW

**Comment by roman-r-m** [comment]
Please could you check if it has any effect on https://github.com/laurent22/joplin/issues/4124

**Comment by laurent22** [comment]
Sorry I don't think we should support this. We add features based on what users need and there has to be a consensus about it and interest in it.

In this case, as far as I know you're the only one who needs it, yet it makes the code significantly more complex (it doesn't help that this part of the code is messy to begin with because it was written before TypeScript and React Hooks). I'm pretty sure there are some edge cases and weird bugs we'll discover after this, and I'm afraid we don't have the resources to deal with this.

**Comment by ken1kob** [comment]
> Sorry I don't think we should support this. We add features based on what users need and there has to be a consensus about it and interest in it.
> 
> In this case, as far as I know you're the only one who needs it, 

I found this topic was discussed several times.
- [Bug: last note doesn't show up on note list - Features - Joplin Forum](https://discourse.joplinapp.org/t/bug-last-note-doesnt-show-up-on-note-list/15241)
- [How to compact note list vertically? - Support - Joplin Forum](https://discourse.joplinapp.org/t/how-to-compact-note-list-vertically/18300)

**Comment by ken1kob** [comment]
> 
> 
> Please could you check if it has any effect on #4124

I confirm the issue for both v.2.1.8 and a build with this fix. The symptom described there never happens in the two versions.

**Comment by ken1kob** [comment]
More topics are found relevant to the height of NoteList items. I cannot say that these are all of them.

- [Sidebar and notelist change size - Support - Joplin Forum](https://discourse.joplinapp.org/t/sidebar-and-notelist-change-size/18410)
- [How reduce vertical white space in note list? - Support - Joplin Forum](https://discourse.joplinapp.org/t/how-reduce-vertical-white-space-in-note-list/4549)
- [How to select certain elements - Lounge - Joplin Forum](https://discourse.joplinapp.org/t/how-to-select-certain-elements/7742/4)

For a long time, many people has been talking about it. But, most of them have given up the feature they wanted. For example, @uxamanda wrote as below in https://discourse.joplinapp.org/t/how-to-select-certain-elements/7742/6

> The selector you’d want is:
> `.note-list > div:not(:first-child):not(:last-child)`
> but changing the height there makes the scroll area short / jittery.

@laurent22, with these evidences, it can be said to be be sufficient to support it, I think.

**Comment by laurent22** [comment]
Indeed you're right that it's been requested before, I hadn't realised.

But also my problem with your implementation is that I have no idea how it works.

For example you set the itemHeight on componentDidUpdate, so after the component renders, but that variable is needed **before** rendering. Yet, somehow, it works.

That itemHeight variable should also be part of the state because it affects rendering, otherwise when you change it in theory nothing should happen. You didn't make it part of the state, yet it works.

So for me there's a bit too much luck involved in any of this working, and I expect some unrelated change anywhere could break the implementation.

I would also like to make this part customisable by plugins and to do that I'll probably refactor the component using React Hook, and converting that logic would be tricky. So I don't know what to suggest. I think ideally the component should be refactored first because adding more fragile code over an already messy component is not great.

**Comment by ken1kob** [comment]
> But also my problem with your implementation is that I have no idea how it works.

My implementation has clear rationales and does not depends on any luck.

The key of the rationales is there are two players, NoteListComponent and NoteListItem. Their lifecycles are different.

The lifetime of NoteListComponent is long, and its itemHeight variable is effective all the while. In single lifetime of NoteListComponent, there are many death and birth events of NoteListeItem instances. NoteListComponent captures the height of one of them and set the value to the itemHeight variable.


> For example you set the itemHeight on componentDidUpdate, so after the component renders, but that variable is needed **before** rendering. Yet, somehow, it works.

There is no magic and uncertainty. The height of NoteListItem specified in CSS is known after the first NoteListItem is mounted. Remember the lifecycles of NoteListComponent and NoteListItem are different. Before the first time of a NoteListComponent is displayed, its componentDidMount() is invoked once, and its componentDidUpdate() is invoked several times, and many NoteListItem instances are created and destroyed. There are many chances to know the height of them (see NoteListComponent.fitItemHeight()).


> That itemHeight variable should also be part of the state because it affects rendering, otherwise when you change it in theory nothing should happen. You didn't make it part of the state, yet it works.

You can put the itemHeight in the state of NoteItemComponent, but it is optional and not mandatory.

The reason why itemHeight can be not in the state is that some members of NoteListComponent's state depend on the itemHeight. Whenever the itemHeight is changed, its dependent state members always be changed. So, putting the itemHeight in the state is not needed.


> So for me there's a bit too much luck involved in any of this working, and I expect some unrelated change anywhere could break the implementation.

My implementation works just as written and does not rely on unknown source codes.

> I would also like to make this part customisable by plugins and to do that I'll probably refactor the component using React Hook, and converting that logic would be tricky. So I don't know what to suggest. I think ideally the component should be refactored first because adding more fragile code over an already messy component is not great.

I agree. To refactor them is good choice.

Since my implementation is small and clear, and it deliver a fair merit for users instantly (More notes shown in search results and a notebook make us happier!!!).

So, until the refactoring starts, how about use this fix? After the refactoring finishes, discard the fix.

**Comment by ken1kob** [comment]
> > For example you set the itemHeight on componentDidUpdate, so after the component renders, but that variable is needed **before** rendering. Yet, somehow, it works.
> 
> Before the first time of a NoteListComponent is displayed, its componentDidMount() is invoked once, and its componentDidUpdate() is invoked several times, 

More precisely speaking, the height of NoteListItem is assigned to the itemHeight variable at a former componentDidUpdate() invocation, and the acquired itemHeight value is used at the latter/final render() invocation.

File: packages/app-desktop/gui/NoteList/NoteList.tsx [hunk 1]
```diff
@@ -425,6 +426,17 @@ class NoteListComponent extends React.Component {
 		}
 	}
 
+	fitItemHeight() {
+		// If a note item are styled by CSS, its height is considered.
+		// NOTE: this.itemHeight doesn't need to be in this.state.
+		const ref = Object.values<any>(this.itemAnchorRefs_)[0];
+		const height = ref && ref.current && ref.current.getHeight();
+		if (height > 0) {
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteList/NoteList.tsx:434]: If the height cannot be determined, shouldn't we set a default (like 34, which is the current item height default) ?

Comment (ken1kob) [reply packages/app-desktop/gui/NoteList/NoteList.tsx:434]: Yes. If height is undefined or null, itemHeight is not touched and its default value (34) remains.

**Comment by deciacco** [comment]
I for one would like to see this implemented and just wanted to say what a great job @ken1kob did to try and get this approved.  Fingers crossed that laurent accepts this or finds another solution. I'm not a big fan of UIs that space things out. I tend to prefer tighter spacing. I also understand the performance aspect, so if there is a way to have both I'm all for it. Thanks.

**Comment by deciacco** [comment]
Been keeping an eye on this... respectfully curious as to why this has not been merged? The fix is there, it is clear, there are people that would like it.

**Comment by placoderm** [comment]
Thanks to everyone working on this issue. I've wanted a fix for this starting the first week I used Joplin. I really appreciate everyone's time and consideration.

**Comment by ken1kob** [comment]
@laurent22, isn't it time to review? Some people (me too) have been waiting for this feature for a long time. 

As I said before, if there is quite some time before NoteList will be refactored using Hook, how about applying this PR for users' immediate convenience?

**Comment by laurent22** [comment]
Sorry I'm not keen to merge this for the reasons given above. I see your point, and I appreciate that there's no other way to implement this, but I also think it's a relatively niche feature.

Additionally I want to stop adding more hacks on top of more hacks. Those parts of the app should be refactored - hopefully it will be done at some point, and in the meantime I prefer if things stay as they are. It's not great that we got to a point where some parts of the app can't be changed in a clean way anymore, but that's just the reality of it. It's the same situation with the sidebar actually, which is another ancient part of the app.

**Comment by placoderm** [comment]
> Additionally I want to stop adding more hacks on top of more hacks.

I'm one of the people who would love this hack, but I completely respect your judgement on not implementing it. I'm sure it's disappointing to the folks who have worked on the proposal, but that's just the way software goes. You're the one responsible to the whole user base so we have to support you. Hopefully there will be support in the future to do the refactoring. I'm sure it's frustrating for you too.

Thanks to you and everyone who works on Joplin (even those folks whose PR's don't get approved)!

**Comment by deciacco** [comment]
This is too bad. ken1kob, spent a good bit of time working on a solution, and a good one at that. Solid enough **to keep the app stable and performing well** - enough to make a good amount of users happy (as he showed) until this can be refactored.  I understand not wanting to add hacks on-top of hacks, **but this would have been just temporary use of good code** until it can be refactored. Again, I would understand a total hack and you can't keep adding hacks on-top of others, but this was good enough to be worth while. Enough so, until there is a better fix, you could have made users happy, including someone that can actually contribute to the project.

**Comment by laurent22** [comment]
> but this would have been just temporary use of good code until it can be refactored.

I don't think you understand how large projects like this work. If we start adding "temporary code" everywhere, which in fact will be there permanently, the whole thing will quickly become unmanageable. That being said, it's open source, so if you can do better you're free to fork the project and improve it on your own.

**Comment by ken1kob** [comment]
Let's go the issue straight.

- Adding a new feature and maintainability are always trade-offs.
- It is agreed this PR is not bad.
- Laurent thinks this feature is "relatively niche".

The point is the last one. If new information that the feature is demanded by many users is found, there is a chance that another decision is thought as better.

Just now, I've made [a vote in the forum](https://discourse.joplinapp.org/t/vote-do-you-want-more-notes-displayed-in-the-note-list/23555). Stay tuned.

**Comment by ken1kob** [comment]
@laurent22, thank you for waiting. The current (2022-02-21) result of the vote is [here](https://discourse.joplinapp.org/t/vote-do-you-want-more-notes-displayed-in-the-note-list/23555/11?u=ken1kob):

image

It shows worthy information as below:
- About 60% people wants more notes displayed. There are some comments with their use cases.
- About 40% people wants selectability/configurability like this PR.
- Of course, this PR satisfies 100% people's needs (both slim and wide). There are no negative comments for  selectability/configurability.
- The amount of the vote is meaningful enough, as far as past votes are concerned.

It can be concluded as below:
- The feature like this PR is fairly demanded by users and **not niche**.
- It is worthy to reconsider the trade-off between adding the feature and maintainability.

**Pull request closed without merging**